### ISSUE TITLE
neo4j-shell "-host" argument not honored/working

### ISSUE THREAD

**Issue by robertklep** [neo4j/neo4j#1432]
Trying to connect to a remote server using Neo4J shell.

Server is Linux (@192.168.23.12) running Neo4j 2.0.0-M06 (Community edition), client is Mac OS X running the same Neo4J version.

This is what I'm trying on the client:

```
$ ./neo4j-shell -host 192.168.23.12 -port 1337 -v
ERROR (-v for expanded information):
    Connection refused
java.rmi.ConnectException: Connection refused to host: 127.0.0.1; nested exception is: 
    java.net.ConnectException: Connection refused
    at sun.rmi.transport.tcp.TCPEndpoint.newSocket(TCPEndpoint.java:619)
    at sun.rmi.transport.tcp.TCPChannel.createConnection(TCPChannel.java:216)
    at sun.rmi.transport.tcp.TCPChannel.newConnection(TCPChannel.java:202)
    at sun.rmi.server.UnicastRef.invoke(UnicastRef.java:128)
        ...
```

(the `127.0.0.1` seems a bit strange...)

I _can_ telnet to 192.168.23.12:1337:

```
$ telnet 192.168.23.12 1337
Trying 192.168.23.12...
Connected to lserver.lan.
Escape character is '^]'.
...
```

I'm running Neo4J in a [Docker](http://www.docker.io/) container on Linux, so I have to explicitly forward ports from the host to the container. I'm forwarding 1337 (as shown by being able to connect using telnet) and 7474. Do I need to forward any other ports maybe?

**Comment by daincredibleholg** [comment]
Hi.

But this is still a flaw in the shell protocol. I opened a question on StackOverflow with the details: http://stackoverflow.com/questions/28974592/strange-neo4j-shell-issue-connecting-to-neo4j-in-a-docker-instance. 
The problem starts once the IP address, you can reach the Neo4j instance from the outside world, isn't visible from the Neo4j instance. The easiest way to explain is your local home network. You usually get one public IP address from your ISP. All your devices in your home network get an internal IP address, e.g. 192.168.1.x. Imagine you run a neo4j server on your home server with the IP 192.168.1.100. This is the only IP address visible for the Neo4j instance. Now, you try to connect from the outside world, say from work, to your Neo4j at home. You use the IP 1.2.3.4. Your home router knows that it has to forward traffic for port 7474 to the internal IP 192.168.1.100. So, the initial incoming connection works. But then, your client at work tries to connect to 192.168.1.100, which is obviously wrong.
Putting the external IP address in the Neo4js server hosts is not really pracitical. To stick with the example: You will get a new external IP address from time to time (or every 24h if you e.g. in Germany). The same is also a problem with the set up, the initial poster described. Your docker container should be as generic as possible, so you never know, which IP address the container will see.
Any chances to get this fixed?

**Comment by rickardoberg** [comment]
@robertklep @daincredibleholg can you try setting "java.rmi.server.hostname" to the IP you want to use in neo4j-wrapper.conf file? That should help.

**Comment by jakewins** [comment]
Since this has not been followed up with the suggested solution above, I'm closing this assuming @rickardoberg's solution solved the issue. Please don't hesitate to re-open this if the issue remains.

**Comment by gorgia** [comment]
This issue persists in 2.3.1 - COMMUNITY with the same behaviour

**Comment by jakewins** [comment]
@Gorgia could you try the suggestion from @rickardoberg above, to help debug this?

**Comment by daincredibleholg** [comment]
Hi. The issue with @rickardoberg's suggestion is: it does not scale. I try to use neo4j in a Docker container and the IP address can (and will) change - esp. if I spin up multiple container. One could suggest to write a start script which is updating the _neo4j-wrapper.conf_, but it is still pretty error-prone. :(

**Comment by jakewins** [comment]
@daincredibleholg agreed, also - the `host` operator is supposed to work, ofc, so this def. seems like a bug. Was just curious if the fix suggested resolves it ish, since then that gives us a good indication of what it is that is wrong. 

In any case, leaving this as an open bug to be picked up by engineering.

**Comment by daincredibleholg** [comment]
Thanks @jakewins. I try to find some spare time to test it. As I am currently in a non-neo4j project, it is a bit tricky to be successful with that search :) I report back if I got some updates.
Anyway, I think the problem is in the basic RMI protocol principle. Using another protocol, maybe a wrapper using the REST interface, could solve the problem (just thinking aloud and I am aware of some downsides of the REST interface approach :))

**Comment by jakewins** [comment]
@daincredibleholg +1 - one avenue to explore is to use the new `Bolt` binary protocol available in Neo4j 3.0. That'd help consolidate things as well.. But it'd be nice to fix this in the interim and for 2.x series releases as well.

**Comment by craigtaverner** [comment]
Neo4j-shell was deprecated in 3.0 in favor of cypher-shell which uses bolt. It would be interesting to know if this problem also occurs with cypher-shell. I think however that neo4j-shell itself will be removed very soon and so this bug is almost certainly not going to be fixed. I will close the issue in the meantime.
